File: src/com/prudentsys/java/agent/DeliverySim.java
package com.prudentsys.agent;

import java.util.*;

public class DeliverySim {
    static final int GRID_SIZE = 6;
    static final int OBSTACLE_COUNT = 8;
    static final int MAX_ENERGY = 20;
    static final int NUM_AGENTS = 2;
    static final double CUSTOMER_SPAWN_CHANCE = 0.3;

    int[][] grid = new int[GRID_SIZE][GRID_SIZE];
    List<int[]> agents = new ArrayList<>();
    List<int[]> customers = new ArrayList<>();
    List<int[]> obstacles = new ArrayList<>();
    List<int[]> agentTargets = new ArrayList<>();
    int[] restaurant = {0, 0};

    int energy = MAX_ENERGY;
    int steps = 0;
    int deliveries = 0;
    int time = 0;
    Random rand = new Random();

    public DeliverySim() {
        for (int i = 0; i < NUM_AGENTS; i++) {
            agents.add(new int[]{restaurant[0], restaurant[1]});
            agentTargets.add(null);
        }
        spawnObstacles();
    }

    void spawnObstacles() {
        while (obstacles.size() < OBSTACLE_COUNT) {
            int x = rand.nextInt(GRID_SIZE);
            int y = rand.nextInt(GRID_SIZE);
            int[] pos = {x, y};
            if (!isSame(pos, restaurant) && !contains(obstacles, pos) && !contains(agents, pos)) {
                obstacles.add(pos);
            }
        }
    }

    void maybeSpawnCustomer() {
        if (rand.nextDouble() < CUSTOMER_SPAWN_CHANCE) {
            while (true) {
                int x = rand.nextInt(GRID_SIZE);
                int y = rand.nextInt(GRID_SIZE);
                int[] pos = {x, y};
                if (!contains(customers, pos) && !isSame(pos, restaurant) && !contains(obstacles, pos)) {
                    customers.add(pos);
                    break;
                }
            }
        }
    }

    void moveAgentToward(int index, int[] target) {
        int[] pos = agents.get(index);
        int ax = pos[0], ay = pos[1];
        int tx = target[0], ty = target[1];

        int[] newPos = Arrays.copyOf(pos, 2);
        if (ax < tx && !isBlocked(ax + 1, ay)) newPos[0]++;
        else if (ax > tx && !isBlocked(ax - 1, ay)) newPos[0]--;
        else if (ay < ty && !isBlocked(ax, ay + 1)) newPos[1]++;
        else if (ay > ty && !isBlocked(ax, ay - 1)) newPos[1]--;

        agents.set(index, newPos);
        steps++;
        energy--;
    }

    boolean isBlocked(int x, int y) {
        int[] pos = {x, y};
        return x < 0 || x >= GRID_SIZE || y < 0 || y >= GRID_SIZE || contains(obstacles, pos) || contains(agents, pos);
    }

    void printGrid() {
        System.out.print("\033[H\033[2J"); // clear console
        System.out.flush();

        for (int i = 0; i < GRID_SIZE; i++) {
            for (int j = 0; j < GRID_SIZE; j++) {
                int[] pos = {i, j};
                if (contains(agents, pos)) System.out.print(" A ");
                else if (isSame(pos, restaurant)) System.out.print(" R ");
                else if (contains(customers, pos)) System.out.print(" C ");
                else if (contains(obstacles, pos)) System.out.print(" # ");
                else System.out.print(" . ");
            }
            System.out.println();
        }
        System.out.printf("⏱️ Time: %d | 🚶 Steps: %d | ✅ Deliveries: %d | ⚡ Energy: %d/%d\n\n",
                time, steps, deliveries, energy, MAX_ENERGY);
    }

    void assignTargets() {
        for (int i = 0; i < NUM_AGENTS; i++) {
            if (agentTargets.get(i) == null || !contains(customers, agentTargets.get(i))) {
                int[] current = agents.get(i);
                int[] closest = null;
                int minDist = Integer.MAX_VALUE;

                for (int[] c : customers) {
                    if (agentTargets.contains(c)) continue;
                    int dist = Math.abs(current[0] - c[0]) + Math.abs(current[1] - c[1]);
                    if (dist < minDist) {
                        minDist = dist;
                        closest = c;
                    }
                }
                agentTargets.set(i, closest);
            }
        }
    }

    void step() {
        time++;
        maybeSpawnCustomer();
        assignTargets();

        for (int i = 0; i < NUM_AGENTS; i++) {
            int[] pos = agents.get(i);
            int[] target = agentTargets.get(i);

            if (energy <= 0 && !isSame(pos, restaurant)) {
                moveAgentToward(i, restaurant);
                continue;
            } else if (isSame(pos, restaurant)) {
                energy = MAX_ENERGY;
            }

            if (target != null) {
                if (isSame(pos, target)) {
                    customers.removeIf(c -> isSame(c, target));
                    agentTargets.set(i, null);
                    deliveries++;
                    System.out.printf("✅ Agent %d delivered to customer at (%d,%d)\n", i, target[0], target[1]);
                } else {
                    moveAgentToward(i, target);
                }
            }
        }
    }

    boolean contains(List<int[]> list, int[] pos) {
        for (int[] p : list) {
            if (isSame(p, pos)) return true;
        }
        return false;
    }

    boolean isSame(int[] a, int[] b) {
        return a[0] == b[0] && a[1] == b[1];
    }

    public static void main(String[] args) throws InterruptedException {
        DeliverySim sim = new DeliverySim();
        while (true) {
            sim.printGrid();
            sim.step();
            Thread.sleep(300);
        }
    }
}

File: src/com/prudentsys/py/agent/fooddelivery/DeliverySim.py
import random
import os
import time

GRID_SIZE = 6
OBSTACLE_COUNT = 8
MAX_ENERGY = 20
NUM_AGENTS = 2
CUSTOMER_SPAWN_CHANCE = 0.3


class DeliverySim:
    def __init__(self):
        self.grid_size = GRID_SIZE
        self.restaurant = [0, 0]
        self.agents = [self.restaurant.copy() for _ in range(NUM_AGENTS)]
        self.agent_targets = [None for _ in range(NUM_AGENTS)]

        self.customers = []
        self.obstacles = self.spawn_obstacles(OBSTACLE_COUNT)

        self.steps_taken = 0
        self.completed_deliveries = 0
        self.time_step = 0
        self.energy = MAX_ENERGY

    def spawn_obstacles(self, count):
        obstacles = []
        while len(obstacles) < count:
            x, y = random.randint(0, GRID_SIZE - 1), random.randint(0, GRID_SIZE - 1)
            if [x, y] != self.restaurant and [x, y] not in self.agents and [x, y] not in obstacles:
                obstacles.append([x, y])
        return obstacles

    def maybe_spawn_customer(self):
        if random.random() < CUSTOMER_SPAWN_CHANCE:
            while True:
                x, y = random.randint(0, GRID_SIZE - 1), random.randint(0, GRID_SIZE - 1)
                if [x, y] not in self.customers and [x, y] != self.restaurant and [x, y] not in self.obstacles:
                    self.customers.append([x, y])
                    break

    def print_grid(self):
        os.system('cls' if os.name == 'nt' else 'clear')
        for i in range(self.grid_size):
            row = ""
            for j in range(self.grid_size):
                cell = [i, j]
                if cell in self.agents:
                    row += " A "
                elif cell == self.restaurant:
                    row += " R "
                elif cell in self.customers:
                    row += " C "
                elif cell in self.obstacles:
                    row += " # "
                else:
                    row += " . "
            print(row)
        print(f"\nTime: {self.time_step} | Steps: {self.steps_taken} | Deliveries: {self.completed_deliveries} | Energy: {self.energy}/{MAX_ENERGY}")

    def assign_customer(self, agent_index):
        if not self.customers:
            return None
        ax, ay = self.agents[agent_index]
        assigned = [c for i, c in enumerate(self.agent_targets) if i != agent_index]
        unassigned = [c for c in self.customers if c not in assigned]
        if not unassigned:
            return None
        return min(unassigned, key=lambda c: abs(ax - c[0]) + abs(ay - c[1]))

    def move_agent_toward(self, index, target):
        ax, ay = self.agents[index]
        tx, ty = target
        new_pos = [ax, ay]

        # Try axis-aligned movement while avoiding obstacles and other agents
        if ax < tx and [ax + 1, ay] not in self.obstacles and [ax + 1, ay] not in self.agents:
            new_pos[0] += 1
        elif ax > tx and [ax - 1, ay] not in self.obstacles and [ax - 1, ay] not in self.agents:
            new_pos[0] -= 1
        elif ay < ty and [ax, ay + 1] not in self.obstacles and [ax, ay + 1] not in self.agents:
            new_pos[1] += 1
        elif ay > ty and [ax, ay - 1] not in self.obstacles and [ax, ay - 1] not in self.agents:
            new_pos[1] -= 1

        self.agents[index] = new_pos

    def step(self):
        self.time_step += 1
        self.maybe_spawn_customer()

        for i in range(NUM_AGENTS):
            pos = self.agents[i]
            target = self.agent_targets[i]

            # Recharge if needed
            if self.energy <= 0 and pos != self.restaurant:
                self.move_agent_toward(i, self.restaurant)
                self.steps_taken += 1
                continue
            elif pos == self.restaurant:
                self.energy = MAX_ENERGY

            if target is None or target not in self.customers:
                target = self.assign_customer(i)
                self.agent_targets[i] = target

            if target:
                if pos == target:
                    self.customers.remove(target)
                    self.agent_targets[i] = None
                    self.completed_deliveries += 1
                    print(f"✅ Agent {i} delivered to customer at {target}!")
                    time.sleep(0.4)
                else:
                    self.move_agent_toward(i, target)
                    self.steps_taken += 1
                    self.energy -= 1

        return True


if __name__ == "__main__":
    sim = DeliverySim()
    while True:
        sim.print_grid()
        if not sim.step():
            break
        time.sleep(0.3)
